kernel/src/callback.rs

```diff
@@ -9,44 +9,122 @@ use crate::process;
+    /// implementing IPC.
+    ///
+    /// This value is crate visible to enable optimizations in sched.rs. Other
+    /// users should call `.index()` instead.
+    crate index: usize,
+
+    /// The unique identifier for this process. This can be used to refer to the
+    /// process in situations where a single number is required, for instance
+    /// when referring to specific applications across the syscall interface.
+    ///
+    /// The combination of (index, identifier) is used to check if the app this
+    /// `AppId` refers to is still valid. If the stored identifier in the
+    /// process at the given index does not match the value saved here, then the
+    /// process moved or otherwise ended, and this `AppId` is no longer valid.
+    identifier: usize,
 }
 
 impl PartialEq for AppId {
     fn eq(&self, other: &AppId) -> bool {
-        self.idx == other.idx
+        self.index == other.index && self.identifier == other.identifier
     }
 }
 
 impl Eq for AppId {}
 
 impl fmt::Debug for AppId {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{}", self.idx)
+        write!(f, "{}", self.identifier)
     }
 }
 
 impl AppId {
-    crate fn new(kernel: &'static Kernel, idx: usize) -> AppId {
+    crate fn new(kernel: &'static Kernel, identifier: usize, index: usize) -> AppId {
         AppId {
             kernel: kernel,
-            idx: idx,
+            identifier: identifier,
+            index: index,
         }
     }
 
-    pub fn idx(&self) -> usize {
-        self.idx
+    /// Get the location of this app in the processes array.
+    ///
+    /// This will return `Some(index)` if the identifier stored in this `AppId`
+    /// matches the app saved at the known index. If the identifier does not
+    /// match then `None` will be returned.
+    crate fn index(&self) -> Option<usize> {
+        // Do a lookup to make sure that the index we have is correct.
+        if self.kernel.appid_is_valid(self) {
+            Some(self.index)
+        } else {
+            None
+        }
+    }
+
+    /// Get a `usize` unique identifier for the app this `AppId` refers to.
+    ///
+    /// This function should not generally be used, instead code should just use
+    /// the `AppId` object itself to refer to various apps on the system.
+    /// However, getting just a `usize` identifier is particularly useful when
+    /// referring to a specific app with things outside of the kernel, say for
+    /// userspace (e.g. IPC) or tockloader (e.g. for debugging) where a concrete
+    /// number is required.
+    ///
+    /// Note, this will always return the saved unique identifier for the app
+    /// originally referred to, even if that app no longer exists. For example,
+    /// the app may have restarted, or may have been ended or removed by the
+    /// kernel. Therefore, calling `id()` is _not_ a valid way to check
+    /// that an application still exists.
```

Comment:
Thinking about global uniqueness of identifiers -- since we're proposing this as a mechanism that labels a process possibly off-core (here tockloader is given as example, but one would also imagine using as an RPC identifier), I think part of the future design discussion for the actual identifiers should include whether we have an entropy source of some kind in the kernel to try to ensure that identifiers are unique even across reboots

Reply:
> [...] whether we have an entropy source of some kind in the kernel to try to ensure that identifiers are unique even across reboots

Note that this would require more bits than `usize` from a cryptographic perspective, if you really want to avoid collision in all realistic scenarios (even considering that embedded CPUs are "slow" enough that one can only generate so many identifiers per second). A collision within a set of random 32-bit identifiers requires [only around 2^16 identifiers](https://en.wikipedia.org/wiki/Birthday_problem).

I don't have much context on the use cases, so I guess it depends whether uniqueness is a *nice to have* property for e.g. debugging, or if a collision is to be avoided at all costs.